File: gamechannel/signatures_tests.cpp
// Copyright (C) 2019-2022 The Xaya developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include "signatures.hpp"

#include "testgame.hpp"

#include <xayautil/base64.hpp>
#include <xayautil/hash.hpp>

#include <gmock/gmock.h>
#include <gtest/gtest.h>

#include <glog/logging.h>

#include <sstream>

namespace xaya
{
namespace
{

using testing::_;
using testing::Return;

class SignaturesTests : public TestGameFixture
{

protected:

  const std::string gameId = "game id";
  const uint256 channelId = SHA256::Hash ("channel id");
  proto::ChannelMetadata meta;

  SignaturesTests ()
  {
    meta.set_reinit (std::string ("re\0init", 7));
    meta.add_participants ()->set_address ("address 0");
    meta.add_participants ()->set_address ("address 1");

    verifier.SetValid ("sgn 0", "address 0");
    verifier.SetValid ("sgn 1", "address 1");
  }

};

TEST_F (SignaturesTests, GetChannelSignatureMessage)
{
  const std::string dataWithNul("foo\0bar", 7);

  std::stringstream expected;
  expected << "Game-Channel Signature\n"
           << "Game ID: " << gameId << "\n"
           << "Channel: " << channelId.ToHex () << "\n"
           << "Reinit: " << EncodeBase64 (std::string ("re\0init", 7)) << "\n"
           << "Topic: topic\n"
           << "Data Hash: " << SHA256::Hash (dataWithNul).ToHex ();

  const std::string actual
      = GetChannelSignatureMessage (gameId, channelId, meta,
                                    "topic", dataWithNul);
  EXPECT_EQ (actual, expected.str ());
  LOG (INFO) << "Signature message:\n" << actual;
}

TEST_F (SignaturesTests, InvalidTopic)
{
  const std::string invalidTopic("a\nb");
  EXPECT_DEATH (GetChannelSignatureMessage (gameId, channelId, meta,
                                            invalidTopic, "foobar"),
                "Topic string contains invalid character");
}

TEST_F (SignaturesTests, VerifyParticipantSignatures)
{
  proto::SignedData data;
  data.set_data ("foobar");
  data.add_signatures ("signature 0");
  data.add_signatures ("signature 1");
  data.add_signatures ("signature 2");

  verifier.ExpectOne (gameId, channelId, meta, "topic", data.data (),
                      "signature 0", "some other address");
  verifier.ExpectOne (gameId, channelId, meta, "topic", data.data (),
                      "signature 1", "address 1");
  verifier.ExpectOne (gameId, channelId, meta, "topic", data.data (),
                      "signature 2", "invalid");

  EXPECT_EQ (VerifyParticipantSignatures (verifier, gameId, channelId, meta,
                                          "topic", data),
             std::set<int> ({1}));
}

TEST_F (SignaturesTests, SignDataForParticipantError)
{
  proto::SignedData data;
  data.set_data ("foobar");
  data.add_signatures ("sgn 0");

  signer.SetAddress ("wrong address");
  EXPECT_FALSE (SignDataForParticipant (signer, gameId, channelId, meta,
                                        "topic", 1, data));

  EXPECT_EQ (VerifyParticipantSignatures (verifier, gameId, channelId, meta,
                                          "topic", data),
             std::set<int> ({0}));
}

TEST_F (SignaturesTests, SignDataForParticipantSuccess)
{
  proto::SignedData data;
  data.set_data ("foobar");
  data.add_signatures ("sgn 0");

  const std::string msg
      = GetChannelSignatureMessage (gameId, channelId, meta,
                                    "topic", data.data ());
  signer.SetAddress ("address 1");
  EXPECT_CALL (signer, SignMessage (msg)).WillOnce (Return ("signature 1"));
  verifier.ExpectOne (gameId, channelId, meta, "topic", data.data (),
                      "signature 1", "address 1");

  ASSERT_TRUE (SignDataForParticipant (signer, gameId, channelId, meta,
                                       "topic", 1, data));

  EXPECT_EQ (VerifyParticipantSignatures (verifier, gameId, channelId, meta,
                                          "topic", data),
             std::set<int> ({0, 1}));
}

} // anonymous namespace
} // namespace xaya

File: gamechannel/signatures_tests.py
#!/usr/bin/env python3

# Copyright (C) 2019-2022 The Xaya developers
# Distributed under the MIT software license, see the accompanying
# file COPYING or http://www.opensource.org/licenses/mit-license.php.

from gamechannel import signatures

from gamechannel.proto import (metadata_pb2, signatures_pb2)

from google.protobuf import text_format

import base64
import unittest
import hashlib


class SignaturesTest (unittest.TestCase):

  def testChannelMessage (self):
    h = hashlib.sha256 ()
    h.update (b"channel id")
    channelId = h.digest ()

    meta = metadata_pb2.ChannelMetadata ()
    meta.reinit = b"re\0init"

    # This is logged by the C++ signatures test.
    msg = """Game-Channel Signature
Game ID: game id
Channel: 537e2fa2e3e5eafcb04ff353ff2a1984b7c1500419d322dcbec8b3613c224d57
Reinit: cmUAaW5pdA==
Topic: topic
Data Hash: d6b681bfce7155d44721afb79c296ef4f0fa80a9dd6b43c5cf74dd0f64c85512"""
    actual = signatures.getChannelMessage ("game id", channelId, meta,
                                           "topic", b"foo\0bar")
    self.assertEqual (actual, msg)

  def testCreateForChannel (self):
    class FakeRpc:
      addresses = ["addr 1", "addr 2"]

      def validateaddress (self, addr):
        return {"isvalid": True}

      def getaddressinfo (self, addr):
        return {"ismine": addr in self.addresses}

      def signmessage (self, addr, msg):
        for i in range (len (self.addresses)):
          if addr == self.addresses[i]:
            return b"sgn %d" % i
        raise AssertionError ("Invalid test address: %s" % addr)

    meta = metadata_pb2.ChannelMetadata ()
    text_format.Parse ("""
      reinit: "reinit"
      participants: { address: "addr 2" }
      participants: { address: "other address" }
    """, meta)

    rpc = FakeRpc ()
    channel = {
      "id": "ab" * 32,
      "meta":
        {
          "proto": base64.b64encode (meta.SerializeToString ()),
        },
    }

    expected = signatures_pb2.SignedData ()
    text_format.Parse ("""
      data: "foobar"
      signatures: "sgn 1"
    """, expected)

    actual = signatures.createForChannel (rpc, "game id", channel,
                                          "topic", b"foobar")
    self.assertEqual (actual, expected)


if __name__ == "__main__":
  unittest.main ()
